README.md
> This file is part of fsio, see <https://qu1x.org/fsio>.
> 
> Copyright (c) 2016 Rouven Spreckels <n3vu0r@qu1x.org>
> 
> fsio is free software: you can redistribute it and/or modify
> it under the terms of the GNU Affero General Public License version 3
> as published by the Free Software Foundation on 19 November 2007.
> 
> fsio is distributed in the hope that it will be useful,
> but WITHOUT ANY WARRANTY; without even the implied warranty of
> MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
> GNU Affero General Public License for more details.
> 
> You should have received a copy of the GNU Affero General Public License
> along with fsio. If not, see <https://www.gnu.org/licenses>.

fsio
====

Feedback Synchronized I/O - CPU/FPGA Communication

Installation
============

Either getting a release
------------------------

1. Download stable source distribution tarball.

		wget https://qu1x.org/file/fsio-1.0.1.tar.xz

2. Extract and enter.

		tar -xJf fsio-1.0.1.tar.xz
		cd fsio-1.0.1

Or getting a snapshot
---------------------

Required dependencies:

		git autoreconf libtool

1. Clone repository.

		git clone https://github.com/qu1x/fsio.git

2. Enter and generate latest source distribution.

		cd fsio
		./bootstrap

Installing one of them
----------------------

Required dependencies:

		build-essential libpugixml

1. Configure, build, and install.

Two features, `--enable-stat` and `--enable-wait`, can be enabled.

		./configure --sysconfdir=/etc
		make
		sudo make install

2. Keep to uninstall someday.

		sudo make uninstall

Usage
=====

RTFM:

		man fsio
		man fsio-tvgen

Have a look at the package definition and the top-level loopback example:

		less fsio.vhd
		less fsio_top.vhd

Communication Protocol
======================

The CPU or processing system (PS) acts as master and requests data to be either
written or read while the FPGA or programmable logic (PL) acts as slave and
serves these requests followed by acknowledging them. The handshake signal as
well as the data signals are fed back in order to synchronize communication,
hence the name Feedback Synchronized I/O (fsio),

The AXI GPIO IP Core from the Xilinx Vivado IP Catalog provides a dual-channel
mode allowing the first channel to be input only and the second channel to be
output only. This suits the protocol well since one of the two channels can be
used as feedback loop reducing the number of maps to not bloat the block design.
An instance of this IP Core is referred to as a map. A communication channel is
made up of a handshake map of one signal and one or more data maps of up to 32
signals.

![Communication Channel](doc/4x4.png)

The mapping is stored in an XML file, a channel description file. It is passed
to the file transfer application named fsio and parsed by the library named
libfsio.

		<fsio>
			<hs i="0x41200000" o="+0x8" width="1"/>
			<fs i="0x41210000" o="+0x8" width="4"/>
			<fs i="0x41220000" o="+0x8" width="4"/>
			<fs i="0x41230000" o="+0x8" width="4"/>
			<fs i="0x41240000" o="+0x8" width="4"/>
		</fsio>

In both figures, the CPU on the left and the FPGA on the right represent an
unidirectional communication channel. The arrows indicate the two feedback loops
of each figure, one for the data and one for the handshake. A loop has both
input and output mappings, labeled “fsi” and “fso” for the data, and “hsi” and
“hso” for the handshake, respectively. What is labeled as input for the CPU, is
labeled as output for the FPGA, and vice versa. Time flows from top to bottom,
except for the orange colored feeding which is done continuously on the FPGA and
alternately with the handshake polling on the CPU.

![Master Write - Slave Read](doc/mwsr.png)

The master write routine firstly writes the data, and repeatedly reads it back
until they are equal. Then it toggles the handshake signal requesting the FPGA
to read the data. Afterwards, the slave read routine of the FPGA acknowledges
the request by feeding back the toggled handshake signal for which the CPU is
waiting for by polling it.

![Master Read - Slave Write](doc/mrsw.png)

In contrast, the master read routine firstly toggles the handshake signal
requesting the FPGA to write data while the CPU repeatedly feeds it back. The
slave write routine of the FPGA compares the written with the fed back data
until they are equal. Afterwards, it acknowledges the request by feeding back
the toggled handshake signal, allowing the CPU to finally read the data.

Performance Measurements
========================

	SoM: Avnet PicoZed 7Z030
	SoC: Zynq XC7Z030 (-1)
	CLK: 180 MHz
	eOS: Xilinx PetaLinux 2015.4
	CFG: --enable-stat
	CXX: arm-linux-gnueabi-gcc -std=c++14 -O3 (4.9.2)
	CMD: fsio -s buf -e 80000000 WIDTH.xml put|get 80MB.tv WIDTH.sv

Write Rate (from CPU to FPGA)
-----------------------------

![Write Rate vs. Channel Width](doc/put-rate.png)

| Maps | Channel Width (B) | Rate (MB/s) | Rate Uncertainty (MB/s) |
| ----:|:----------------- | -----------:|:----------------------- |
|    1 | 4                 |       4.540 | 0.006                   |
|    2 | 8                 |       6.288 | 0.016                   |
|    3 | 12                |       7.231 | 0.008                   |
|    4 | 16                |       7.798 | 0.015                   |
|    5 | 20                |       8.095 | 0.038                   |
|    6 | 24                |       8.452 | 0.009                   |
|    7 | 28                |       8.672 | 0.010                   |
|    8 | 32                |       8.827 | 0.010                   |
|    9 | 36                |       8.986 | 0.016                   |
|   10 | 40                |       9.095 | 0.013                   |
|   11 | 44                |       9.135 | 0.012                   |
|   12 | 48                |       9.203 | 0.015                   |
|   13 | 52                |       9.314 | 0.013                   |
|   14 | 56                |       9.386 | 0.013                   |
|   15 | 60                |       9.452 | 0.013                   |
|   16 | 64                |       9.472 | 0.014                   |
|   17 | 68                |       9.461 | 0.015                   |
|   18 | 72                |       9.480 | 0.015                   |
|   19 | 76                |       9.555 | 0.018                   |
|   20 | 80                |       9.594 | 0.019                   |
|   21 | 84                |       9.630 | 0.017                   |
|   22 | 88                |       9.652 | 0.017                   |
|   23 | 92                |       9.679 | 0.017                   |
|   24 | 96                |       9.687 | 0.018                   |
|   25 | 100               |       9.717 | 0.018                   |
|   26 | 104               |       9.733 | 0.018                   |
|   27 | 108               |       9.737 | 0.019                   |
|   28 | 112               |       9.734 | 0.019                   |
|   29 | 116               |       9.735 | 0.019                   |
|   30 | 120               |       9.734 | 0.020                   |
|   31 | 124               |       9.736 | 0.020                   |
|   32 | 128               |       9.733 | 0.021                   |

Read Rate (from FPGA to CPU)
----------------------------

![Read Rate vs. Channel Width](doc/get-rate.png)

| Maps | Channel Width (B) | Rate (MB/s) | Rate Uncertainty (MB/s) |
| ----:|:----------------- | -----------:|:----------------------- |
|    1 | 4                 |       0.855 | 0.057                   |
|    2 | 8                 |       1.567 | 0.091                   |
|    3 | 12                |       2.184 | 0.114                   |
|    4 | 16                |       2.722 | 0.123                   |
|    5 | 20                |       3.191 | 0.131                   |
|    6 | 24                |       3.605 | 0.130                   |
|    7 | 28                |       3.979 | 0.126                   |
|    8 | 32                |       4.337 | 0.118                   |
|    9 | 36                |       4.611 | 0.110                   |
|   10 | 40                |       4.894 | 0.102                   |
|   11 | 44                |       5.052 | 0.086                   |
|   12 | 48                |       5.319 | 0.077                   |
|   13 | 52                |       5.507 | 0.063                   |
|   14 | 56                |       5.676 | 0.058                   |
|   15 | 60                |       5.862 | 0.042                   |
|   16 | 64                |       6.049 | 0.030                   |
|   17 | 68                |       6.167 | 0.020                   |
|   18 | 72                |       6.276 | 0.009                   |
|   19 | 76                |       6.405 | 0.008                   |
|   20 | 80                |       6.534 | 0.015                   |
|   21 | 84                |       6.626 | 0.022                   |
|   22 | 88                |       6.738 | 0.026                   |
|   23 | 92                |       6.848 | 0.030                   |
|   24 | 96                |       6.958 | 0.031                   |
|   25 | 100               |       7.017 | 0.031                   |
|   26 | 104               |       7.092 | 0.032                   |
|   27 | 108               |       7.195 | 0.030                   |
|   28 | 112               |       7.266 | 0.027                   |
|   29 | 116               |       7.335 | 0.027                   |
|   30 | 120               |       7.385 | 0.025                   |
|   31 | 124               |       7.438 | 0.025                   |
|   32 | 128               |       7.545 | 0.021                   |

Version
=======

fsio-1.0.1 <https://qu1x.org/fsio>

License
=======

GNU Affero General Public License version 3

Authors
=======

* Copyright (c) 2016 Rouven Spreckels <n3vu0r@qu1x.org>



NEWS
2016-06-09

 * 1.0.1: Added documentation and performance measurements.

2016-05-03

 * 1.0.0: Initial release

# This file is part of fsio, see <https://qu1x.org/fsio>.
# 
# Copyright (c) 2016 Rouven Spreckels <n3vu0r@qu1x.org>
# 
# fsio is free software: you can redistribute it and/or modify
# it under the terms of the GNU Affero General Public License version 3
# as published by the Free Software Foundation on 19 November 2007.
# 
# fsio is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
# GNU Affero General Public License for more details.
# 
# You should have received a copy of the GNU Affero General Public License
# along with fsio. If not, see <https://www.gnu.org/licenses>.


src/lib/fsio.cxx
// This file is part of libfsio, see <https://qu1x.org/libfsio>.
// 
// Copyright (c) 2016 Rouven Spreckels <n3vu0r@qu1x.org>
// 
// libfsio is free software: you can redistribute it and/or modify
// it under the terms of the GNU Affero General Public License version 3
// as published by the Free Software Foundation on 19 November 2007.
// 
// libfsio is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
// GNU Affero General Public License for more details.
// 
// You should have received a copy of the GNU Affero General Public License
// along with libfsio. If not, see <https://www.gnu.org/licenses>.

#include "fsio.hxx"
#include "fsio_map.hxx"

#include <fstream>
#include <string>
#include <cstdint>
#include <cstdlib>
#include <cstring>
#include <cerrno>

#include <sys/types.h>
#include <sys/stat.h>
#include <fcntl.h>
#include <unistd.h>
#include <time.h>

std::string fsio::text(std::string file) {
	std::string dst;
	try {
		std::ifstream src(file.data(), std::ios::in | std::ios::binary);
		src.exceptions(std::ifstream::failbit | std::ifstream::badbit);
		src.seekg(0, std::ios::end);
		dst.resize(src.tellg());
		src.seekg(0, std::ios::beg);
		src.read(&dst.front(), dst.size());
	}
	catch (std::ifstream::failure&) {
		throw "Cannot read file " + file + ": " + std::strerror(errno);
	}
	return dst;
}

void fsio::load(pugi::xml_document& tree, std::string file, std::string root) {
	auto xml = text(file);
	auto res = tree.load(xml.data());
	if (!res) {
		std::string err(res.description());
		if (res.offset) {
			std::string msg(xml.data() + res.offset);
			msg = msg.substr(0, msg.find_first_of('\n') - 1);
			err += " at character " + std::to_string(res.offset)
				+ " [..." + msg + "...]";
		}
		throw err;
	}
	if (root.size() && !tree.child(root.data()))
		throw "Missing root node " + root;
}

auto fsio::make(int fd, pugi::xml_node node) -> std::unique_ptr<map> {
	std::size_t i = std::strtoul(node.attribute("i").value(), nullptr, 0);
	std::size_t o = std::strtoul(node.attribute("o").value(), nullptr, 0);
	bool d = node.attribute("o").value()[0] == '+';
	switch (node.attribute("width").as_uint()) {
		case 1: return std::make_unique<fsio_map<std::uint8_t>>(fd, i, o, d);
		case 2: return std::make_unique<fsio_map<std::uint16_t>>(fd, i, o, d);
		case 4: return std::make_unique<fsio_map<std::uint32_t>>(fd, i, o, d);
		case 8: return std::make_unique<fsio_map<std::uint64_t>>(fd, i, o, d);
		default: throw std::string("Unsupported map width");
	}
}

void fsio::make(pugi::xml_node node) {
	int fd = open("/dev/mem", O_RDWR | O_SYNC);
	if (fd == -1)
		throw std::string("Cannot open memory: ") + std::strerror(errno);
	hs = make(fd, node.child("hs"));
	for (auto item: node.children("fs"))
		fs.emplace_back(make(fd, item));
	if (close(fd) == -1)
		throw std::string("Cannot close memory: ") + std::strerror(errno);
	fs_size = 0;
	for (auto& io: fs)
		fs_size += io->size();
}

fsio::fsio(pugi::xml_node node, bool feed, bool back, std::size_t poll):
fsio(feed, back, poll) {
	make(node);
}

fsio::fsio(std::string file, bool feed, bool back, std::size_t poll):
fsio(feed, back, poll) {
	pugi::xml_document tree;
	load(tree, file);
	make(tree.first_child());
}

#if FSIO_WAIT
void fsio::wait() const {
	if (poll) {
		timespec req = { 0, (long)poll };
		nanosleep(&req, 0);
	}
}
#endif

void fsio::feed() const {
	auto feed = [this] () {
		for (auto& io: fs)
			io->feed();
	};
	for (hs->wave(), feed(); !hs->back(); wait())
		feed();
}

void fsio::back() const {
	for (bool back = false; !back; wait()) {
		back = true;
		for (auto& io: fs)
			if (!io->back())
				back = false;
	}
	wave();
}

void fsio::stat(std::vector<std::size_t>& sv) const {
	sv.clear();
	sv.reserve(fs.size() + 1);
	sv.push_back(hs->stat());
	for (auto& io: fs)
		sv.push_back(io->stat());
}

auto fsio::get(void* data) const -> std::size_t {
	feed_data ? feed() : wave();
	for (auto& io: fs)
		data = (char*)data + io->get(data);
	return fs_size;
}

auto fsio::put(const void* data) const -> std::size_t {
	for (auto& io: fs)
		data = (const char*)data + io->put(data);
	back_data ? back() : wave();
	return fs_size;
}


src/lib/fsio.vhd
-- This file is part of fsio, see <https://qu1x.org/fsio>.
-- 
-- Copyright (c) 2016 Rouven Spreckels <n3vu0r@qu1x.org>
-- 
-- fsio is free software: you can redistribute it and/or modify
-- it under the terms of the GNU Affero General Public License version 3
-- as published by the Free Software Foundation on 19 November 2007.
-- 
-- fsio is distributed in the hope that it will be useful,
-- but WITHOUT ANY WARRANTY; without even the implied warranty of
-- MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
-- GNU Affero General Public License for more details.
-- 
-- You should have received a copy of the GNU Affero General Public License
-- along with fsio. If not, see <https://www.gnu.org/licenses>.

library ieee;
use ieee.std_logic_1164.all;

package fsio is
	constant CAP: integer := 32;
	constant LEN: integer := CAP;
	component fsio_get is
		generic (
			-- Number of signals of all data maps
			cap: integer := CAP;
			-- Number of signals of all data maps actually being used
			len: integer := LEN
		);
		port (
			-- AXI core clock
			clk: in std_logic;
			-- AXI handshake input
			hsi: in std_logic;
			-- AXI handshake output as conditional feedback
			hso: out std_logic;
			-- AXI data input of all data maps
			fsi: in std_logic_vector(cap - 1 downto 0);
			-- AXI data output as feedback of all data maps
			fso: out std_logic_vector(cap - 1 downto 0);
			-- User data of all data maps to be read
			dat: out std_logic_vector(len - 1 downto 0);
			-- PS requests PL to read the data
			-- Is set until acknowledged and the data has been fed back
			req: out std_logic;
			-- PL acknowledges when it has read the data
			-- Must be kept set until request has been reset
			ack: in std_logic
		);
	end component fsio_get;
	component fsio_put is
		generic (
			-- Number of signals of all data maps
			cap: integer := CAP;
			-- Number of signals of all data maps actually being used
			len: integer := LEN
		);
		port (
			-- AXI core clock
			clk: in std_logic;
			-- AXI handshake input
			hsi: in std_logic;
			-- AXI handshake output as conditional feedback
			hso: out std_logic;
			-- AXI data input as feedback of all data maps
			fsi: in std_logic_vector(cap - 1 downto 0);
			-- AXI data output of all data maps
			fso: out std_logic_vector(cap - 1 downto 0);
			-- User data of all data maps to be written
			dat: in std_logic_vector(len - 1 downto 0);
			-- PS requests PL to write the data
			-- Is set until acknowledged
			req: out std_logic;
			-- PL acknowledges when it has written the data
			-- Must be set for exactly one clock cycle
			ack: in std_logic
		);
	end component fsio_put;
end package;


src/lib/fsio_get.vhd
-- This file is part of fsio, see <https://qu1x.org/fsio>.
-- 
-- Copyright (c) 2016 Rouven Spreckels <n3vu0r@qu1x.org>
-- 
-- fsio is free software: you can redistribute it and/or modify
-- it under the terms of the GNU Affero General Public License version 3
-- as published by the Free Software Foundation on 19 November 2007.
-- 
-- fsio is distributed in the hope that it will be useful,
-- but WITHOUT ANY WARRANTY; without even the implied warranty of
-- MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
-- GNU Affero General Public License for more details.
-- 
-- You should have received a copy of the GNU Affero General Public License
-- along with fsio. If not, see <https://www.gnu.org/licenses>.

library ieee;
use ieee.std_logic_1164.all;

library fsio;
use fsio.fsio.all;

entity fsio_get is
	generic (
		cap: integer := CAP;
		len: integer := LEN
	);
	port (
		clk: in std_logic;
		hsi: in std_logic;
		hso: out std_logic;
		fsi: in std_logic_vector(cap - 1 downto 0);
		fso: out std_logic_vector(cap - 1 downto 0);
		dat: out std_logic_vector(len - 1 downto 0);
		req: out std_logic;
		ack: in std_logic
	);
end fsio_get;

architecture behavioral of fsio_get is
begin
	dat <= fso(len - 1 downto 0);
	req <= hso xor hsi;
	fso <= fsi;
	ctl: process(clk)
	begin
		if rising_edge(clk) then
			hso <= hso xor (req and ack);
		end if;
	end process ctl;
end behavioral;


src/lib/fsio_put.vhd
-- This file is part of fsio, see <https://qu1x.org/fsio>.
-- 
-- Copyright (c) 2016 Rouven Spreckels <n3vu0r@qu1x.org>
-- 
-- fsio is free software: you can redistribute it and/or modify
-- it under the terms of the GNU Affero General Public License version 3
-- as published by the Free Software Foundation on 19 November 2007.
-- 
-- fsio is distributed in the hope that it will be useful,
-- but WITHOUT ANY WARRANTY; without even the implied warranty of
-- MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
-- GNU Affero General Public License for more details.
-- 
-- You should have received a copy of the GNU Affero General Public License
-- along with fsio. If not, see <https://www.gnu.org/licenses>.

library ieee;
use ieee.std_logic_1164.all;

library fsio;
use fsio.fsio.all;

entity fsio_put is
	generic (
		cap: integer := CAP;
		len: integer := LEN
	);
	port (
		clk: in std_logic;
		hsi: in std_logic;
		hso: out std_logic;
		fsi: in std_logic_vector(cap - 1 downto 0);
		fso: out std_logic_vector(cap - 1 downto 0);
		dat: in std_logic_vector(len - 1 downto 0);
		req: out std_logic;
		ack: in std_logic
	);
end fsio_put;

architecture behavioral of fsio_put is
begin
	fso(len - 1 downto 0) <= dat;
	req <= hso xor hsi;
	ctl: process(clk)
	begin
		if rising_edge(clk) then
			if fso = fsi then
				hso <= hso xor (req and ack);
			end if;
		end if;
	end process ctl;
end behavioral;


src/lib/fsio_top.vhd
-- This file is part of fsio, see <https://qu1x.org/fsio>.
-- 
-- Copyright (c) 2016 Rouven Spreckels <n3vu0r@qu1x.org>
-- 
-- fsio is free software: you can redistribute it and/or modify
-- it under the terms of the GNU Affero General Public License version 3
-- as published by the Free Software Foundation on 19 November 2007.
-- 
-- fsio is distributed in the hope that it will be useful,
-- but WITHOUT ANY WARRANTY; without even the implied warranty of
-- MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
-- GNU Affero General Public License for more details.
-- 
-- You should have received a copy of the GNU Affero General Public License
-- along with fsio. If not, see <https://www.gnu.org/licenses>.

library ieee;
use ieee.std_logic_1164.all;
use ieee.numeric_std.all;

library fsio;
use fsio.fsio.all;

entity fsio_top is
	port (
		DDR_addr : inout std_logic_vector(14 downto 0);
		DDR_ba : inout std_logic_vector(2 downto 0);
		DDR_cas_n : inout std_logic;
		DDR_ck_n : inout std_logic;
		DDR_ck_p : inout std_logic;
		DDR_cke : inout std_logic;
		DDR_cs_n : inout std_logic;
		DDR_dm : inout std_logic_vector(3 downto 0);
		DDR_dq : inout std_logic_vector(31 downto 0);
		DDR_dqs_n : inout std_logic_vector(3 downto 0);
		DDR_dqs_p : inout std_logic_vector(3 downto 0);
		DDR_odt : inout std_logic;
		DDR_ras_n : inout std_logic;
		DDR_reset_n : inout std_logic;
		DDR_we_n : inout std_logic;
		FIXED_IO_ddr_vrn : inout std_logic;
		FIXED_IO_ddr_vrp : inout std_logic;
		FIXED_IO_mio : inout std_logic_vector(53 downto 0);
		FIXED_IO_ps_clk : inout std_logic;
		FIXED_IO_ps_porb : inout std_logic;
		FIXED_IO_ps_srstb : inout std_logic
	);
end fsio_top;

architecture behavioral of fsio_top is
	component ps is
		port (
			DDR_cas_n : inout std_logic;
			DDR_cke : inout std_logic;
			DDR_ck_n : inout std_logic;
			DDR_ck_p : inout std_logic;
			DDR_cs_n : inout std_logic;
			DDR_reset_n : inout std_logic;
			DDR_odt : inout std_logic;
			DDR_ras_n : inout std_logic;
			DDR_we_n : inout std_logic;
			DDR_ba : inout std_logic_vector(2 downto 0);
			DDR_addr : inout std_logic_vector(14 downto 0);
			DDR_dm : inout std_logic_vector(3 downto 0);
			DDR_dq : inout std_logic_vector(31 downto 0);
			DDR_dqs_n : inout std_logic_vector(3 downto 0);
			DDR_dqs_p : inout std_logic_vector(3 downto 0);
			FIXED_IO_mio : inout std_logic_vector(53 downto 0);
			FIXED_IO_ddr_vrn : inout std_logic;
			FIXED_IO_ddr_vrp : inout std_logic;
			FIXED_IO_ps_srstb : inout std_logic;
			FIXED_IO_ps_clk : inout std_logic;
			FIXED_IO_ps_porb : inout std_logic;

			clk180 : out std_logic;

			hsi0 : in std_logic_vector(0 to 0);
			hso0 : out std_logic_vector(0 to 0);

			fsi1 : in std_logic_vector(31 downto 0);
			fso1 : out std_logic_vector(31 downto 0);
			fsi2 : in std_logic_vector(31 downto 0);
			fso2 : out std_logic_vector(31 downto 0);
			fsi3 : in std_logic_vector(31 downto 0);
			fso3 : out std_logic_vector(31 downto 0);
			fsi4 : in std_logic_vector(31 downto 0);
			fso4 : out std_logic_vector(31 downto 0);

			hsi5 : in std_logic_vector(0 to 0);
			hso5 : out std_logic_vector(0 to 0);

			fsi6 : in std_logic_vector(31 downto 0);
			fso6 : out std_logic_vector(31 downto 0);
			fsi7 : in std_logic_vector(31 downto 0);
			fso7 : out std_logic_vector(31 downto 0);
			fsi8 : in std_logic_vector(31 downto 0);
			fso8 : out std_logic_vector(31 downto 0);
			fsi9 : in std_logic_vector(31 downto 0);
			fso9 : out std_logic_vector(31 downto 0)
		);
	end component ps;

	signal clk180: std_logic;

	signal req0, ack0: std_logic;
	signal hsi0, hso0: std_logic;
	signal fsi0, fso0, dat0: std_logic_vector(127 downto 0);

	signal req1, ack1: std_logic;
	signal hsi1, hso1: std_logic;
	signal fsi1, fso1, dat1: std_logic_vector(127 downto 0);

	signal data: std_logic_vector(127 downto 0);
	signal full: std_logic;
begin
	ps0: component ps
	port map (
		DDR_addr(14 downto 0) => DDR_addr(14 downto 0),
		DDR_ba(2 downto 0) => DDR_ba(2 downto 0),
		DDR_cas_n => DDR_cas_n,
		DDR_ck_n => DDR_ck_n,
		DDR_ck_p => DDR_ck_p,
		DDR_cke => DDR_cke,
		DDR_cs_n => DDR_cs_n,
		DDR_dm(3 downto 0) => DDR_dm(3 downto 0),
		DDR_dq(31 downto 0) => DDR_dq(31 downto 0),
		DDR_dqs_n(3 downto 0) => DDR_dqs_n(3 downto 0),
		DDR_dqs_p(3 downto 0) => DDR_dqs_p(3 downto 0),
		DDR_odt => DDR_odt,
		DDR_ras_n => DDR_ras_n,
		DDR_reset_n => DDR_reset_n,
		DDR_we_n => DDR_we_n,
		FIXED_IO_ddr_vrn => FIXED_IO_ddr_vrn,
		FIXED_IO_ddr_vrp => FIXED_IO_ddr_vrp,
		FIXED_IO_mio(53 downto 0) => FIXED_IO_mio(53 downto 0),
		FIXED_IO_ps_clk => FIXED_IO_ps_clk,
		FIXED_IO_ps_porb => FIXED_IO_ps_porb,
		FIXED_IO_ps_srstb => FIXED_IO_ps_srstb,

		clk180 => clk180,

		hsi0(0) => hso0,
		hso0(0) => hsi0,

		fsi1(31 downto 0) => fso0(31 downto 0),
		fso1(31 downto 0) => fsi0(31 downto 0),
		fsi2(31 downto 0) => fso0(63 downto 32),
		fso2(31 downto 0) => fsi0(63 downto 32),
		fsi3(31 downto 0) => fso0(95 downto 64),
		fso3(31 downto 0) => fsi0(95 downto 64),
		fsi4(31 downto 0) => fso0(127 downto 96),
		fso4(31 downto 0) => fsi0(127 downto 96),

		hsi5(0) => hso1,
		hso5(0) => hsi1,

		fsi6(31 downto 0) => fso1(31 downto 0),
		fso6(31 downto 0) => fsi1(31 downto 0),
		fsi7(31 downto 0) => fso1(63 downto 32),
		fso7(31 downto 0) => fsi1(63 downto 32),
		fsi8(31 downto 0) => fso1(95 downto 64),
		fso8(31 downto 0) => fsi1(95 downto 64),
		fsi9(31 downto 0) => fso1(127 downto 96),
		fso9(31 downto 0) => fsi1(127 downto 96)
	);

	fsio0: fsio_put
	generic map (
		cap => 128,
		len => 128
	)
	port map (
		clk => clk180,
		hsi => hsi0,
		hso => hso0,
		fsi => fsi0,
		fso => fso0,
		dat => dat0,
		req => req0,
		ack => ack0
	);

	fsio1: fsio_get
	generic map (
		cap => 128,
		len => 128
	)
	port map (
		clk => clk180,
		hsi => hsi1,
		hso => hso1,
		fsi => fsi1,
		fso => fso1,
		dat => dat1,
		req => req1,
		ack => ack1
	);

	ctl: process(clk180)
	begin
		if rising_edge(clk180) then
			if req0 then
				if full and not ack0 then
					dat0 <= data;
					full <= '0';
					ack0 <= '1';
				end if;
			else
				ack0 <= '0';
			end if;

			ack1 <= '0';
			if req1 then
				if not full then
					data <= dat1;
					full <= '1';
					ack1 <= '1';
				end if;
			end if;
		end if;
	end process ctl;
end behavioral;
